File: HW12/HW12.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;



// // There are n houses in a village. We want to supply water for all the houses by building wells and laying pipes.

// For each house i, we can either build a well inside it directly with cost wells[i - 1] (note the -1 due to 0-indexing), or pipe in water from another well to it. The costs to lay pipes between houses are given by the array pipes where each pipes[j] = [house1j, house2j, costj] represents the cost to connect house1j and house2j together using a pipe. Connections are bidirectional, and there could be multiple valid connections between the same two houses with different costs.

// Return the minimum total cost to supply water to all houses.

/**
  2 <= n <= 104
  wells.length == n
 0 <= wells[i] <= 105
  1 <= pipes.length <= 104
 pipes[j].length == 3
 1 <= house1j, house2j <= n
 0 <= costj <= 105
 house1j != house2j
 */

public class HW12 {
//dont need this function since im not doing prims anymore
    public static int findMin(int n, Boolean contains [], ArrayList<int []> pipes)
    {
        int min = Integer.MAX_VALUE;
        for(int i = 0; i < n; i++)
        {
            if(contains[i] == false && pipes.get(i)[2] < min)
            {
                min = i;
            }
        }

        return min;
    }
//dont need this little shitter either 
    public static Boolean checkwells(int val, int [] wells, int [] pipeEdge)
    {
        if(val-wells[val] < pipeEdge[2] )
        {
            return true;
        }
        return false;
    }

    public static void mergeSort(ArrayList<int[]> edges) {
        if (edges == null || edges.size() <= 1) {
            return;
        }
        mergeSortHelper(edges, 0, edges.size() - 1);
    }

    private static void mergeSortHelper(ArrayList<int[]> edges, int start, int end) {
        if (start < end) {
            int mid = start + (end - start) / 2;
            mergeSortHelper(edges, start, mid);
            mergeSortHelper(edges, mid + 1, end);
            merge(edges, start, mid, end);
        }
    }

    private static void merge(ArrayList<int[]> edges, int start, int mid, int end) {
        int n1 = mid - start + 1;
        int n2 = end - mid;

        int[][] left = new int[n1][];
        int[][] right = new int[n2][];

        for (int i = 0; i < n1; ++i) {
            left[i] = edges.get(start + i);
        }
        for (int j = 0; j < n2; ++j) {
            right[j] = edges.get(mid + 1 + j);
        }

        int i = 0, j = 0, k = start;

        while (i < n1 && j < n2) {
            if (left[i][2] <= right[j][2]) {
                edges.set(k, left[i]);
                i++;
            } else {
                edges.set(k, right[j]);
                j++;
            }
            k++;
        }

        while (i < n1) {
            edges.set(k, left[i]);
            i++;
            k++;
        }

        while (j < n2) {
            edges.set(k, right[j]);
            j++;
            k++;
        }
    }

       
    
    public static int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {
        // Your code here
        ArrayList<int []> edges = new ArrayList<>(); //holds edges in arraylist instead of arr -- not rerally needed but done for convenience
        
        for(int pipe []: pipes){
            edges.add(pipe);
        }

        for(int i = 0; i < n; i++)
        {
            edges.add(new int[] {0, i+1, wells[i]});
        }

        mergeSort(edges);

        int [] MST = new int [n+1];

        for(int i = 0; i <= n; i++)
        {
            MST[i] = i;
        }

        int count = 0;
        int cost = 0;

        for (int [] edge: edges)
        {
            int from = edge[0];
            int to = edge[1];
            int forCost = edge[2];
            
            if(find(MST, from) != find(MST, to))
            {
                union(MST, from, to);
                cost += forCost;
                count++;

                if(count == n){
                     break;
                }

            
            }
        }

        return cost;




     }

        // Find for finding the set of the node 'i'
    private static int find(int[] parent, int i) {
        if (parent[i] != i) {
            parent[i] = find(parent, parent[i]);
        }
        return parent[i];
    }
    
    // Union for adding an edge to the MST
    private static void union(int[] parent, int i, int j) {
        int rootI = find(parent, i);
        int rootJ = find(parent, j);
        if (rootI != rootJ) {
            parent[rootI] = rootJ;
        }
    }



     public static void main(String[] args)
     {
        int[] wells1 = {1,1};
        int[][] pipes1 = {{1,2,1},{1,2,2}};
        System.out.println(minCostToSupplyWater(2, wells1, pipes1)); // Should be 2

        // Test Case 3: Pipe is cheaper
        int[] wells2 = {10, 20, 30};
        int[][] pipes2 = {{1, 2, 5}, {2, 3, 5}, {1, 3, 20}};
        System.out.println(minCostToSupplyWater(3, wells2, pipes2)); // Should 20

        // Test Case 4: Cheap pipe tiebreak
        int[] wells3 = {50, 50, 50, 50};
        int[][] pipes3 = {{1, 2, 10}, {2, 3, 10}, {3, 4, 10}, {1, 3, 30}};
        System.out.println(minCostToSupplyWater(4, wells3, pipes3)); // Should be 80
     }
}

File: HW12/HW12.py
from typing import List

def find(parent, i):
    if parent[i] != i:
        parent[i] = find(parent, parent[i])
    return parent[i]

def union(parent, i, j):
    root_i = find(parent, i)
    root_j = find(parent, j)
    if root_i != root_j:
        parent[root_i] = root_j

def merge_sort(edges):
    if len(edges) <= 1:
        return
    mid = len(edges) // 2
    left_half = edges[:mid]
    right_half = edges[mid:]
    merge_sort(left_half)
    merge_sort(right_half)

    i = j = k = 0
    while i < len(left_half) and j < len(right_half):
        if left_half[i][2] <= right_half[j][2]:
            edges[k] = left_half[i]
            i += 1
        else:
            edges[k] = right_half[j]
            j += 1
        k += 1

    while i < len(left_half):
        edges[k] = left_half[i]
        i += 1
        k += 1

    while j < len(right_half):
        edges[k] = right_half[j]
        j += 1
        k += 1

def minCostToSupplyWater(self, n: int, wells: List[int], pipes: List[List[int]]) -> int:
    edges = []
    for pipe in pipes:
        edges.append(pipe)
    for i in range(n):
        edges.append([0, i + 1, wells[i]])

    merge_sort(edges)

    MST = list(range(n + 1))

    count = 0
    cost = 0

    for edge in edges:
        from_, to, for_cost = edge
        if find(MST, from_) != find(MST, to):
            union(MST, from_, to)
            cost += for_cost
            count += 1
            if count == n:
                break

    return cost

# Test cases
wells1 = [1, 1]
pipes1 = [[1, 2, 1], [1, 2, 2]]
print(minCostToSupplyWater(None, 2, wells1, pipes1))  # Should be 2

wells2 = [10, 20, 30]
pipes2 = [[1, 2, 5], [2, 3, 5], [1, 3, 20]]
print(minCostToSupplyWater(None, 3, wells2, pipes2))  # Should be 20

wells3 = [50, 50, 50, 50]
pipes3 = [[1, 2, 10], [2, 3, 10], [3, 4, 10], [1, 3, 30]]
print(minCostToSupplyWater(None, 4, wells3, pipes3))  # Should be 80
